### PR TITLE
Update impact, rigid groundcheck, moving platform

### PR DESCRIPTION
**Pull request by nezvers**
Added Ray collision shape to these tests.
RigidBody ground check has added floor segment with convex shape, previously only original concave. I noticed that on the original concave floor in Bullet sphere and ray didn't lose ground but on concave they started to flicker. Also added floor segmentation should give extra test points.
The moving platform now has a movement up and down. It will be needed to test the ground connection (Bullet KinematicBody connection is unstable on vertical movement).

### CODE REVIEW

File: 3d/physics_tests/tests/functional/test_moving_platform.gd [hunk 1]
```diff
@@ -10,6 +10,13 @@ const OPTION_FRICTION = "Physics options/Friction (Rigid only)"
 const OPTION_ROUGH = "Physics options/Rough (Rigid only)"
 const OPTION_PROCESS_PHYSICS = "Physics options/AnimationPlayer physics process mode"
 
+const SHAPE_CAPSULE = "Collision shapes/Capsule"
+const SHAPE_BOX = "Collision shapes/Box"
+const SHAPE_CILINDER = "Collision shapes/Cilinder"
```
Comment (pouleyKetchoupp) [inline 3d/physics_tests/tests/functional/test_moving_platform.gd:15]: Typo: `CYLINDER` instead of `CILINDER` (in the scene as well).

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Looks great!

For the ground check test, it would be nice to be able to switch between concave and convex floor with an option menu instead of having to wait for the objects to move all the way through one then the other, for faster testing.

File: 3d/physics_tests/tests/functional/test_moving_platform.gd (hunk 1 above)
Comment (nezvers) [reply 3d/physics_tests/tests/functional/test_moving_platform.gd]: Done.

**Comment by nezvers** [comment]
Now includes options menu for switching floor sizes (small, big) and collision shape (Convex, concave, box).

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn [hunk 2]
```diff
@@ -62,98 +62,294 @@ friction = 0.0
...
+collision_layer = 2
+collision_mask = 3
+
+[node name="MeshInstance" type="MeshInstance" parent="Floors/ConvexSmall/ConvexFloor2"]
+visible = false
+mesh = SubResource( 23 )
+material/0 = null
+
+[node name="CollisionShape" type="CollisionShape" parent="Floors/ConvexSmall/ConvexFloor2"]
+visible = false
```
[the remark this answers is not in the document]
Comment (nezvers) [reply 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn]: Done.

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn [hunk 3]
```diff
@@ -62,98 +62,294 @@ friction = 0.0
...
+[node name="MeshInstance" type="MeshInstance" parent="Floors/BoxBig/BoxFloor"]
+visible = false
+mesh = SubResource( 23 )
+material/0 = null
+
+[node name="CollisionShape" type="CollisionShape" parent="Floors/BoxBig/BoxFloor"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, -1, 0 )
+shape = SubResource( 30 )
+
 [node name="Camera" type="Camera" parent="."]
```
[the remark this answers is not in the document]
Comment (nezvers) [reply 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn:348]: Done.

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.gd [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+var _current_floor_name = SHAPE_CONCAVE
+var _current_bodies
+var _current_floor
+
+
+func _ready():
+	var options = $Options
+	_dynamic_shapes_scene = get_packed_scene($DynamicShapes/Bodies)
+	_floor_shapes[SHAPE_CONVEX+"Small"] = get_packed_scene($"Floors/ConvexSmall")
```
[the remark this answers is not in the document]
Comment (nezvers) [reply 3d/physics_tests/tests/functional/test_rigidbody_ground_check.gd]: Done.

**Review by pouleyKetchoupp — APPROVED** [review]
Looks good!

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn [hunk 5]
```diff
@@ -62,98 +62,294 @@ friction = 0.0
...
+[node name="RigidBodyRay" type="RigidBody" parent="DynamicShapes/Bodies"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 8.42391, 0, 0 )
+physics_material_override = SubResource( 17 )
+axis_lock_angular_x = true
+axis_lock_angular_y = true
+axis_lock_angular_z = true
+script = ExtResource( 1 )
+
+[node name="CollisionShape" type="CollisionShape" parent="DynamicShapes/Bodies/RigidBodyRay"]
+transform = Transform( 0.8, 0, 0, 0, -3.49691e-08, -0.8, 0, 0.8, -3.49691e-08, 0, 0, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn:207]:
```suggestion
transform = Transform( 0.8, 0, 0, 0, 0, -0.8, 0, 0.8, 0, 0, 0, 0 )
```

File: 3d/physics_tests/tests/functional/test_moving_platform.tscn [hunk 6]
```diff
@@ -40,43 +66,85 @@ __meta__ = {
...
+[node name="Sphere" type="CollisionShape" parent="Bodies/KinematicBody"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0.8, 0 )
+shape = SubResource( 4 )
+
+[node name="Convex" type="CollisionShape" parent="Bodies/KinematicBody"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0.8, 0 )
+shape = SubResource( 5 )
+
+[node name="Ray" type="CollisionShape" parent="Bodies/KinematicBody"]
+transform = Transform( 1, 0, 0, 0, -4.37114e-08, -1, 0, 1, -4.37114e-08, 0, 0.8, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_moving_platform.tscn:102]:
```suggestion
transform = Transform( 1, 0, 0, 0, 0, -1, 0, 1, 0, 0, 0.8, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 7]
```diff
@@ -55,7 +58,7 @@ transform = Transform( 0.8, 0, 0, 0, 1, 0, 0, 0, 0.8, 0, 0, 0 )
 shape = SubResource( 3 )
 
 [node name="RigidBodyConvex" type="RigidBody" parent="DynamicShapes"]
-transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 3, 0.974548, 0 )
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 3, 1.97795, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:61]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 3, 2, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 8]
```diff
@@ -34,7 +36,7 @@ transform = Transform( 0.6, 0, 0, 0, 1, 0, 0, 0, 0.6, 0, 0, 0 )
 shape = SubResource( 1 )
 
 [node name="RigidBodyCapsule" type="RigidBody" parent="DynamicShapes"]
-transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -3, 0, 0 )
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -3, 1.0034, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:39]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -3, 1, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 9]
```diff
@@ -45,6 +47,7 @@ transform = Transform( 0.8, 0, 0, 0, -3.49691e-08, -0.8, 0, 0.8, -3.49691e-08, 0
 shape = SubResource( 2 )
 
 [node name="RigidBodyCylinder" type="RigidBody" parent="DynamicShapes"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1.0034, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:50]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 10]
```diff
@@ -16,14 +16,16 @@ points = PoolVector3Array( -0.7, 0, -0.7, -0.3, 0, 0.8, 0.8, 0, -0.3, 0, -1, 0 )
...
+
 [node name="Test" type="Spatial"]
 script = ExtResource( 2 )
 
 [node name="DynamicShapes" type="Spatial" parent="."]
 transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 9.35591, 0 )
 
 [node name="RigidBodyBox" type="RigidBody" parent="DynamicShapes"]
-transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -6, 0, 0 )
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -6, 1.0034, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:28]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -6, 1, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 11]
```diff
@@ -66,7 +69,7 @@ transform = Transform( 1.5, 0, 0, 0, 2, 0, 0, 0, 1.5, 0, 0, 0 )
 shape = SubResource( 4 )
 
 [node name="RigidBodySphere" type="RigidBody" parent="DynamicShapes"]
-transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 6, 0, 0 )
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 6, 1.0034, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:72]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 6, 1, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 12]
```diff
@@ -76,6 +79,17 @@ script = ExtResource( 1 )
 transform = Transform( 0.8, 0, 0, 0, 0.8, 0, 0, 0, 0.8, 0, 0, 0 )
 shape = SubResource( 5 )
 
+[node name="RigidBodyRay" type="RigidBody" parent="DynamicShapes"]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 8, 1.0034, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:83]:
```suggestion
transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 8, 1, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn [hunk 13]
```diff
@@ -76,6 +79,17 @@ script = ExtResource( 1 )
...
+axis_lock_angular_x = true
+axis_lock_angular_y = true
+axis_lock_angular_z = true
+script = ExtResource( 1 )
+
+[node name="CollisionShape" type="CollisionShape" parent="DynamicShapes/RigidBodyRay"]
+transform = Transform( 0.8, 0, 0, 0, -3.49691e-08, -0.8, 0, 0.8, -3.49691e-08, 0, 0, 0 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_impact.tscn:90]:
```suggestion
transform = Transform( 0.8, 0, 0, 0, 0, -0.8, 0, 0.8, 0, 0, 0, 0 )
```

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn [hunk 14]
```diff
@@ -62,98 +62,294 @@ friction = 0.0
...
+shape = SubResource( 21 )
+
+[node name="MeshInstance" type="MeshInstance" parent="DynamicShapes/Bodies/RigidBodyRay/CollisionShape"]
+mesh = SubResource( 22 )
+material/0 = SubResource( 20 )
+
+[node name="Floors" type="Spatial" parent="."]
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1 )
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_ground_check.tscn:215]:
```suggestion
```

This doesn't need to be translated (and with the translation gone, the transform is the identity so it's not stored).

**Review by aaronfranke — APPROVED** [review]
Overall this looks very good!

Why are the collision shapes scaled by `0.8`? Is it to test scaled collision shapes? It's generally discouraged to scale physics stuff in Godot, but if it's for testing purposes it 100% makes sense.

I left some nit-picky suggestions below. I set this as approved because even if you don't act on them I will still merge this as this PR is very good.

File: 3d/physics_tests/tests/functional/test_rigidbody_ground_check.gd [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+	if _current_bodies:
+		_current_bodies.queue_free()
+	if _current_floor:
+		_current_floor.queue_free()
+
+	var dynamic_bodies = _dynamic_shapes_scene.instance()
+	_current_bodies = dynamic_bodies
+	add_child(dynamic_bodies)
+
+	var floor_inst = _floor_shapes[_current_floor_name+_floor_size].instance()
```
Comment (aaronfranke) [inline 3d/physics_tests/tests/functional/test_rigidbody_ground_check.gd:64]:
```suggestion
	var floor_inst = _floor_shapes[_current_floor_name + _floor_size].instance()
```

File: 3d/physics_tests/tests/functional/test_moving_platform.tscn [hunk 16]
```diff
@@ -40,43 +66,85 @@ __meta__ = {
...
+
+[node name="Ray" type="CollisionShape" parent="Bodies/RigidBody"]
+transform = Transform( 1, 0, 0, 0, -4.37114e-08, -1, 0, 1, -4.37114e-08, 0, 0.8, 0 )
+shape = SubResource( 6 )
+
 [node name="Platforms" type="Spatial" parent="."]
 
 [node name="KinematicPlatform" type="KinematicBody" parent="Platforms"]
-transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -7, -2, 0 )
+transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, -7, -4.23538, 0 )
```
[the remark this answers is not in the document]
Comment (nezvers) [reply 3d/physics_tests/tests/functional/test_moving_platform.tscn:140]: It's a moving platform, while moving it will go all kinds of float numbers.

**Review by aaronfranke — APPROVED** [review]

**Pull request merged**